controllers/hostingautoscalingpolicy/hostingautoscalingpolicy_controller.go

```diff
@@ -0,0 +1,450 @@
+	var err error
+
+	// Ensure we are using the job name specified in the spec
+	if ctx.HostingAutoscalingPolicy.Spec.PolicyName != nil && len(*ctx.HostingAutoscalingPolicy.Spec.PolicyName) > 0 {
+		ctx.PolicyName = *ctx.HostingAutoscalingPolicy.Spec.PolicyName
+	} else {
+		ctx.PolicyName = DefaultAutoscalingPolicyName
+		ctx.HostingAutoscalingPolicy.Spec.PolicyName = &ctx.PolicyName
+
+		// Review: Is this needed
+		if err := r.Update(ctx, ctx.HostingAutoscalingPolicy); err != nil {
+			ctx.Log.Info("Error while updating HostingAutoscalingPolicy policyName in spec")
+			return err
+		}
+	}
+
+	// Save the ResourceIDs into ctx as usable strings
+	if err = r.getResourceIDListfromInputSpec(ctx); err != nil {
+		ctx.Log.Error(err, "Error reading the ResourceIDs from Spec")
+		return err
+	}
+
+	// Initialize other values to defaults if not in Spec
+	if ctx.HostingAutoscalingPolicy.Spec.ScalableDimension == nil {
+		namespace := sdkutil.HostingAutoscalingPolicyServiceNamespace
+		ctx.HostingAutoscalingPolicy.Spec.ServiceNamespace = &namespace
+	}
+
+	if ctx.HostingAutoscalingPolicy.Spec.ScalableDimension == nil {
+		dimension := ScalableDimension
+		ctx.HostingAutoscalingPolicy.Spec.ScalableDimension = &dimension
+	}
+
+	if ctx.HostingAutoscalingPolicy.Spec.PolicyType == nil {
+		policyType := PolicyType
+		ctx.HostingAutoscalingPolicy.Spec.PolicyType = &policyType
+	}
+
+	awsConfig, err := r.awsConfigLoader.LoadAwsConfigWithOverrides(*ctx.HostingAutoscalingPolicy.Spec.Region, ctx.HostingAutoscalingPolicy.Spec.SageMakerEndpoint)
+	if err != nil {
+		ctx.Log.Error(err, "Error loading AWS config")
+		return err
+	}
+
+	ctx.ApplicationAutoscalingClient = r.createApplicationAutoscalingClient(awsConfig)
+	ctx.Log.Info("Loaded AWS config")
+
+	return nil
+}
+
+// determineActionForAutoscaling checks if controller needs to create/delete/update HAP
+func (r *Reconciler) determineActionForAutoscaling(ctx reconcileRequestContext) (controllers.ReconcileAction, error) {
+	var err error
+	// Review: Should you add a check based on status too, for controller restarts
+	if controllers.HasDeletionTimestamp(ctx.HostingAutoscalingPolicy.ObjectMeta) {
+		ctx.Log.Info("Object Has Deletion Timestamp")
+		if len(ctx.ScalingPolicyDescriptionList) > 0 {
+			return controllers.NeedsDelete, nil
+		}
+		return controllers.NeedsNoop, nil
+	}
+
+	//Review: Second condition is not needed, cleanup after test
+	if ctx.ScalingPolicyDescriptionList == nil || len(ctx.ScalingPolicyDescriptionList) == 0 {
+		return controllers.NeedsCreate, nil
+	}
+
+	var comparison sdkutil.Comparison
+
+	if comparison, err = sdkutil.HostingAutoscalingPolicySpecMatchesDescription(ctx.ScalableTargetDescriptionList, ctx.ScalingPolicyDescriptionList, ctx.HostingAutoscalingPolicy.Spec); err != nil {
+		return controllers.NeedsNoop, err
+	}
+
+	r.Log.Info("Compared existing HAP to actual HAP to determine if HAP needs to be updated.", "differences", comparison.Differences, "is-equal", comparison.Equal)
```

Comment:
This could be spammy?

Reply:
Is also in the model and could help debug update errors since it prints the diff.